File: stack/stack_ll.go


package main

import (
	"fmt"
)

type stack_data struct {
     
	data int;
    next *stack_data

}


type  stack struct {

	data *stack_data 
}


func newStack () (*stack) {

	st := new(stack)
	st.data = nil

	return st

}

func (st *stack) push (element int) () {

	if (st == nil) {
		fmt.Println ("stack not initialize")
		return 
	}

	temp := new (stack_data)

	temp.data = element
	temp.next = nil

	if st.data == nil  {

		st.data  = temp
	    
	} else {
        
		temp.next = st.data
		st.data = temp

	}	

}

func (st *stack) pop () (int) {

	if (st == nil) {
		fmt.Println ("stack not initialize")
		return  -1
	}

	element  := -1

	if st.data == nil {

		fmt.Println ("empty stack")

	} else {

	   element = st.data.data 
	   st.data = st.data.next
	}   

	return element

}

func main () {

	fmt.Println ("stack using linklist")

	var st *stack = newStack () 

	st.push (10)

	fmt.Println (st.pop ())
	fmt.Println (st.pop ())

}

File: stack/stack_ll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct stack_data_t {
    struct stack_data_t *next; 
    int data;
}stack_data;

typedef struct stack_t {  
    
    stack_data *data;
} stack;

void 
initilalize_stack (stack **st) {

    stack *temp = NULL;

    if ((*st) == NULL) {

        temp = ( stack *)malloc (sizeof (stack) );
        if (temp == NULL) {
            printf ("No memory \n");
        } else {

            temp->data = NULL;          
            *st = temp;
        }
     }

       
}

void 
push (stack **st , int element) {

     
    if ((*st) == NULL) {
         printf ("Stack not initialize \n");
    }
    
    stack_data *temp = NULL;
    temp = (stack_data *) malloc (sizeof (stack_data ));

    temp->data = element;
    temp->next = NULL;

    if (((*st)->data) == NULL) {
        /* First element */
        (*st)->data = temp;       
    } else {
        temp->next = (*st)->data;
        (*st)->data = temp;
    }
}

int 
pop (stack **st ) {

    int element = -1;
     
    if ((*st) == NULL) {
         printf ("Stack not initialize \n");
    }


    if (((*st)->data) == NULL) {
        printf ("stack empty \n");      
    } else {
        element  = (*st)->data->data;
        (*st)->data = (*st)->data->next;
    }

    return element;
}


int 
main (void) {

    printf ("Stack implementing with linklist \n");

    stack *st = NULL;

    initilalize_stack (&st);

    push (&st, 10);

    printf ("%d \n",pop(&st));
     printf ("%d \n",pop(&st));

}